strawberryfields/api_client.py

```diff
@@ -0,0 +1,584 @@
+
+        config = {k: v for k, v in self.DEFAULT_CONFIG.items()}
+
+        # Try getting everything first from configuration
+        config.update(self.get_configuration_from_config())
+
+        # Override any values that are explicitly passed when initializing client
+        config.update(kwargs)
+
+        if config["hostname"] is None:
+            raise ValueError("hostname parameter is missing")
+
+        if config["hostname"] not in self.ALLOWED_HOSTNAMES:
+            raise ValueError("hostname parameter not in allowed list")
+
+        self.USE_SSL = config["use_ssl"]
+        if not self.USE_SSL:
+            warnings.warn("Connecting insecurely to API server", UserWarning)
+
+        self.HOSTNAME = config["hostname"]
+        self.BASE_URL = "{}://{}".format("https" if self.USE_SSL else "http", self.HOSTNAME)
+        self.AUTHENTICATION_TOKEN = config["authentication_token"]
+        self.HEADERS = {"User-Agent": self.USER_AGENT}
+
+        if self.AUTHENTICATION_TOKEN is not None:
+            self.set_authorization_header(self.AUTHENTICATION_TOKEN)
+
+    def get_configuration_from_config(self):
+        """
+        Retrieve configuration from environment variables or config file based on Strawberry Fields
+        configuration.
+        """
+        return configuration.Configuration().api
+
+    def authenticate(self, username, password):
+        """
+        Retrieve an authentication token from the server via username
+        and password authentication and calls set_authorization_header.
+
+        Args:
+            username (str): A user name.
+            password (str): A password.
+        """
+        raise NotImplementedError()
+
+    def set_authorization_header(self, authentication_token):
+        """
+        Adds the authorization header to the headers dictionary to be included
+        with all API requests.
+
+        Args:
+            authentication_token (str): An authentication token used to access the API.
+        """
+        self.HEADERS["Authorization"] = authentication_token
+
+    def join_path(self, path):
+        """
+        Joins a base url with an additional path (e.g., a resource name and ID)
+
+        Args:
+            path (str): A path to be joined with BASE_URL.
+
+        Returns:
+            str: A joined path.
+        """
+        return join_path(self.BASE_URL, path)
+
+    def get(self, path):
+        """
+        Sends a GET request to the provided path. Returns a response object.
+
+        Args:
+            path (str): A path to send the GET request to.
+
+        Returns:
+            requests.Response: A response object, or None if no response could be fetched from the
+            server.
+        """
+        try:
+            response = requests.get(url=self.join_path(path), headers=self.HEADERS)
+        except requests.exceptions.ConnectionError as e:
+            response = None
+            warnings.warn("Could not connect to server ({})".format(e))
```

Comment:
Rather than a warning, should this be a `ConnectionError`?

Reply:
@josh146 this has come up before -- upon more thinking, I think it does make sense to leave this as an exception and let the user decide what to do with it. Will change. However, there is a separate ticket for better handling and messaging of errors.

Reply:
Done